server/src-lib/Hasura/Server/CacheUpdate.hs

```diff
@@ -0,0 +1,178 @@
+  , cacheUpdateEventListener
+  , cacheUpdateEventProcessor
+  )
+where
+
+import           Hasura.Prelude
+
+import           Hasura.Logging
+import           Hasura.RQL.DDL.Schema.Table (buildSchemaCache)
+import           Hasura.RQL.Types
+import           Hasura.Server.App           (withLock)
+import           Hasura.Server.Query
+import           Hasura.Server.Utils         (bsToTxt)
+
+import           Control.Concurrent.MVar
+import           Data.Aeson
+import           Data.Aeson.Casing
+import           Data.Aeson.TH
+import           Data.UUID
+import           Database.PG.Query
+
+import qualified Control.Concurrent          as C
+import qualified Control.Concurrent.STM      as STM
+import qualified Data.IORef                  as IORef
+import qualified Data.Text                   as T
+import qualified Data.Time                   as UTC
+import qualified Database.PostgreSQL.LibPQ   as PQ
+import qualified Network.HTTP.Client         as HTTP
+
+pgChannel :: PGChannel
+pgChannel = "hasura_cache_update"
+
+data ThreadType
+  = TTListener
+  | TTProcessor
+  deriving (Eq)
+
+instance Show ThreadType where
+  show TTListener  = "listener"
+  show TTProcessor = "processor"
+
+
+data CacheUpdateEventLog
+  = CacheUpdateEventLog
+  { cuelLogLevel   :: !LogLevel
+  , cuelThreadType :: !ThreadType
+  , cuelInfo       :: !Value
+  } deriving (Show, Eq)
+
+instance ToJSON CacheUpdateEventLog where
+  toJSON (CacheUpdateEventLog _ t info) =
+    object [ "thread_type" .= show t
+           , "info" .= info
+           ]
+
+instance ToEngineLog CacheUpdateEventLog where
+  toEngineLog threadLog =
+    (cuelLogLevel threadLog, "cache_update_event", toJSON threadLog)
+
+data EventPayload
+  = EventPayload
+  { _epServerId   :: !UUID
+  , _epOccurredAt :: !UTC.UTCTime
+  } deriving (Show, Eq)
+$(deriveJSON (aesonDrop 3 snakeCase) ''EventPayload)
+
+data CacheUpdateEvent
+  = CUEListenSuccess !EventPayload
+  | CUEListenFail
+  deriving (Show, Eq)
+
+instance ToJSON CacheUpdateEvent where
+  toJSON (CUEListenSuccess payload) = toJSON payload
+  toJSON CUEListenFail              = String "event listening failed"
+
+data ThreadError
+  = TEJsonParse !T.Text
+  | TEQueryError !QErr
+$(deriveToJSON
+  defaultOptions { constructorTagModifier = snakeCase . drop 2
+                 , sumEncoding = TaggedObject "type" "info"
+                 }
+ ''ThreadError)
+
+-- | An IO action that listens to postgres for events and pushes them to a Queue
+cacheUpdateEventListener
+  :: PGPool
+  -> Logger
+  -> STM.TQueue CacheUpdateEvent
+  -> IO ()
+cacheUpdateEventListener pool logger eventsQueue =
+  -- Never exits
+  forever $ do
+    listenResE <- liftIO $ runExceptT $ listen pool pgChannel notifyHandler
+    either onError return listenResE
+  where
+    threadType = TTListener
+
+    onError e = do
+      logError logger threadType $ TEQueryError e
+      -- Push a listen failed event to queue
+      STM.atomically $ STM.writeTQueue eventsQueue CUEListenFail
```

Comment:
Don't emit a 'listen error' event. Emit an event once the connection is re-established after an error. You'll probably need to expose this functionality from pg-client-hs, probably `listen` can take something like this as an argument:
```haskell
data NotifyHandler 
  = NotifyHandler 
  { _ehOnInit :: IO ()
  , _ehOnMessage :: Notify -> IO ()
  }
```